### PR TITLE
Python 3 is coming

### PR DESCRIPTION
**Pull request by rochacbruno**
https://pythonclock.org/

### CODE REVIEW

**Comment by JacobCallahan** [comment]
Is this part of a larger effort to move completely away from relying on six?

**Comment by rochacbruno** [comment]
@JacobCallahan we are going to keep six for a while, the idea here is to fix some incompatibilities that Travis is not able to find and run some test cases in Jenkins CI until we have it all working and move the master to Python 3.6.

@kbidarkar started to look what is needed on Jenkins side as it can be difficult to setup Python3.6 binaries there.

**Comment by kbidarkar** [comment]
@rochacbruno and @renzon , I was looking into moving to Python3 and below are my findings.

1) Currently added Python3.5 to the list of python binaries on the jenkins.

2) We should be able to run with Python 3.5 readily , but for Python 3.6 , we would require an OS update on all the slaves.

3) If Python3.6 is something that is must we would need to update the OS first, only then we would be able to run jobs on it. 

4) The current jenkins-slaves of sat6-rhel6 and sat6-rhel7 both are on Fedora25 and don't have access available readily to make use of Python3.6, but have Python3.5 readily available.

5) Unless something is a hard dependency on Python3.6, I would suggest:
a) We move first to Python3 by making use of Python3.5 binary / Fedora25
b) Test out all the jobs for a while with Python3/Python3.5 
c) If all looks good with jobs running with Python3/Python3.5   we shall go ahead and update the jenkins slave with latest Fedora OS to have access to Python3.6

Let me know, if the plan seems ok ?

6) Probably we can try running the Satellite6 Installer with Python3.5 sometime soon if we have a branch  ready with requirements.txt changes. Feel working with `Satellite6 Installer`  would be a good place to start testing out things and slowly update and accomodate stuff required for the move towards Python3/Python3.5
```
[root@jenkins-slave ~]# ll /usr/bin/python3
lrwxrwxrwx. 1 root root 9 Oct  9 08:09 /usr/bin/python3 -> python3.5
[root@jenkins-slave ~]# cat /etc/redhat-release 
Fedora release 25 (Twenty Five)
[root@jenkins-slave ~]# rpm -qf /usr/bin/python3.5
python3-3.5.4-2.fc25.x86_64
[root@jenkins-slave ~]# dnf update python3
Last metadata expiration check: 1:24:41 ago on Mon Nov 20 10:31:16 2017.
Dependencies resolved.
Nothing to do.
Complete!
[root@jenkins-slave ~]# 
```

**Comment by kbidarkar** [comment]
Also having Python3.6 on Jenkins Master is not a hard requirement as we don't run any jobs on that.
If the jenkins-slaves are configured to have the Python3.5 or Python3.6 binaries we are good to move towards Python3 for running the jobs.

**Comment by renzon** [comment]
The main feature we would loose running Python 3.5 instead of 3.6 is f strings. Besides that they changed also dicts to be consume less memory, mainly when creating classes. If it's not difficult to add it, I would go with 3.6 ;)

**Comment by rochacbruno** [comment]
Hi @kbidarkar as @omaciel  mentioned on mtg, we are going to move to Python 3 only next year,
It is important to have Python 3.6 so we can use new features such as f-strings.

Let's get Python 3.6 setup by the start of the year? 

Let me know how can I help you with this task!

:)

**Comment by rochacbruno** [comment]
Pyup says testimony, inflector and unitest2 is not python3 ready, we need to check it.

@elyezer do you know something about testimony? is it Py3 compatible?

screenshot_2017-12-22_13-28-31

File: robottelo/ssh.py [hunk 1]
```diff
@@ -22,8 +22,10 @@ class SSHCommandTimeoutError(Exception):
 
 
 def decode_to_utf8(text):  # pragma: no cover
-    """In python 3 all strings are already unicode, no need to decode"""
-    if six.PY2:
+    """Paramiko returns bytes object and we need to ensure it is utf-8 before
+    parsing
+    """
+    if isinstance(text, bytes):
```
Comment (elyezer) [inline robottelo/ssh.py]: I would suggest using EAFP (https://docs.python.org/3/glossary.html#term-eafp), having in mind that:

```python
>>> 'str'.decode
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'str' object has no attribute 'decode'
>>> b'bytes'.decode
<built-in method decode of bytes object at 0x7faa1f2236e8>
```

Than you could use as:

```python
try:
    return text.decode('utf-8')
except AttributeError:
    return text
```

**Comment by elyezer** [comment]
@rochacbruno testimony should run on python3. If you run robotello on python3 and run make to validate docstrings then you are running testimony on python3 too. I've always run testimony on python3 and had no issues so far.

Need to check why pyup is thinking testimony is not python3 compatible to fix that.

**Comment by kbidarkar** [comment]
@rochacbruno, Once `sesame`, `paprika` is updated to the latest Fedora 27 OS, running Python3.6 should be straightforward. 

I will be on PTO, from Feb10th to Feb 25th, will be back after that.

In the meantime, if anyone wants to upgrade `paprika` or `sesame`, please do so, after taking backup of the network files on it.

**Comment by abalakh** [comment]
Addressed @elyezer suggestion. Results without the fix:
```python
stderr = regex.sub('', decode_to_utf8(stderr))
TypeError: cannot use a string pattern on a bytes-like object
```
Results with fix applied for test with ssh commands, vms etc:
```python
pytest -v tests/foreman/ui_airgun/test_activationkey.py -k test_positive_delete_with_system
============================= test session starts ==============================
platform darwin -- Python 3.6.4, pytest-3.3.2, py-1.5.2, pluggy-0.6.0 -- /Users/andrii/workspace/py3a/bin/python3.6
cachedir: .cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: wait-for-1.0.9, xdist-1.22.0, services-1.2.1, mock-1.6.3, forked-0.2
collected 5 items
2018-03-20 17:59:57 - conftest - DEBUG - BZ deselect is disabled in settings


tests/foreman/ui_airgun/test_activationkey.py::test_positive_delete_with_system PASSED [100%]

============================== 4 tests deselected ==============================
=================== 1 passed, 4 deselected in 330.20 seconds ===================
```

**Comment by abalakh** [comment]
Rebased the branch to include all the changes for ~0.5 year :) And especially #5858 . @kbidarkar mind retriggering automation against py3 branch?

File: robottelo/ssh.py [hunk 2]
```diff
@@ -22,10 +22,13 @@ class SSHCommandTimeoutError(Exception):
...
-    """In python 3 all strings are already unicode, no need to decode"""
-    if six.PY2:
+    """Paramiko returns bytes object and we need to ensure it is utf-8 before
+    parsing
+    """
+    try:
         return text.decode('utf-8')
-    return text
+    except AttributeError:
+        return text
```
Comment (ldjebran) [inline robottelo/ssh.py]: Tghis should work for python 2 and python 3
```python
def decode_to_utf8(text): 
    try:
        return text.decode('utf-8')
    except (AttributeError, UnicodeEncodeError):
        return text
```

**Comment by kbidarkar** [comment]
@SatelliteQE/collaborators, we now have sat6.3.z branch and this PR is raised against `master` branch, so we can now go ahead and merge this PR.

Also, we have multiple folks reviewed and ACKed this PR, will merge this now.

**Comment by kbidarkar** [comment]
@SatelliteQE/collaborators, we now have the 6.3.z support PR branch here #5860 
I will cherry-pick and add all the commits to this PR.

Will wait for sat6.2.z jobs also work with py3, before merging this PR, this is being done to simplify changes needed to get Jenkins jobs working for both.
Let's move both sat6.2.z and sat6.3.z jobs together to py3 to avoid job creation complexity separately with separate py3 and py2.7 interpreters.

The plan is to keep, sat6.3.z and sat6.2.z py3 PR's ready with any py3  fixes ready and merge after Sat6.3.1 and sat6.2.15 releases.

**Pull request merged**